Applicant Name - Suteja Patil

Email id- patil.sut@northeastern.edu

###Question 1: Given some sample data, write a program to answer the following: click here to access the required data set

On Shopify, we have exactly 100 sneaker shops, and each of these shops sells only one model of shoe. We want to do some analysis of the average order value (AOV). When we look at orders data over a 30 day window, we naively calculate an AOV of $3145.13. Given that we know these shops are selling sneakers, a relatively affordable item, something seems wrong with our analysis. 



###Q1a- Think about what could be going wrong with our calculation. Think about a better way to evaluate this data. 


In [27]:
import os
import pandas as pd

In [28]:
orders = pd.read_csv("2019 Winter Data Science Intern Challenge Data Set.csv")

In [29]:
orders

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0            1       53      746           224            2           cash   
1            2       92      925            90            1           cash   
2            3       44      861           144            1           cash   
3            4       18      935           156            1    credit_card   
4            5       18      883           156            1    credit_card   
...        ...      ...      ...           ...          ...            ...   
4995      4996       73      993           330            2          debit   
4996      4997       48      789           234            2           cash   
4997      4998       56      867           351            3           cash   
4998      4999       60      825           354            2    credit_card   
4999      5000       44      734           288            2          debit   

               created_at  
0     2017-03-13 12:36:56  
1     2017-03-03 17:38:52  
2      2017-03-14 4:23:56  
3     2017-03-26 12:43:37  
4      2017-03-01 4:35:11  
...                   ...  
4995  2017-03-30 13:47:17  
4996  2017-03-16 20:36:16  
4997   2017-03-19 5:42:42  
4998  2017-03-16 14:51:18  
4999  2017-03-18 15:48:18  

[5000 rows x 7 columns]

In [30]:
orders.describe()

order_id      shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720
std    1443.520003    29.006118    87.798982   41282.539349    116.32032
min       1.000000     1.000000   607.000000      90.000000      1.00000
25%    1250.750000    24.000000   775.000000     163.000000      1.00000
50%    2500.500000    50.000000   849.000000     284.000000      2.00000
75%    3750.250000    75.000000   925.000000     390.000000      3.00000
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000

In [31]:
orders['order_amount'].median()

284.0

###Answer 1a - The mean of order_amount is 3145 but median is 284. This means that there are some entries that is skewing the mean. So the problem with the given calculation is that outliers have not been handled. If outliers are handled, we will get a more generalized and better AOV.So we can handle outliers and use median to calculate the correct AOV .

###Q1b - What metric would you report for this dataset?


###Answer -  The data is only over a 30 day window. So the data is limited and we cant estimate about the diversity in future data. So the correct metric for this dataset for now will be AOV. But since outliers have not been removed yet, we will need to pre-process the data to remove outliers and calculate the correct AOV.



In [32]:
orders['total_items'].value_counts().sort_index()

1       1830
2       1832
3        941
4        293
5         77
6          9
8          1
2000      17
Name: total_items, dtype: int64

We can see above that most people have ordered only one or two pair of sneakers while soma have ordered 2000. So this large number of order like 2000 which is an outlier is skewing the mean.
So to handle this I am simply going to remove the entries with total_orders value that is 1.5 times more than the calculated inter quartile range.



In [33]:
q1 = orders["total_items"].quantile(q=0.25)
q3 = orders["total_items"].quantile(q=0.75)

IQR = q3-q1
IQR

2.0

In [34]:
q2 = orders["total_items"].quantile(q=0.5)

orders_df = orders[(orders["total_items"] < (q2 + IQR * 1.5)) & 
                   (orders["total_items"] > (q2 - IQR * 1.5))]

In [35]:
orders_df

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0            1       53      746           224            2           cash   
1            2       92      925            90            1           cash   
2            3       44      861           144            1           cash   
3            4       18      935           156            1    credit_card   
4            5       18      883           156            1    credit_card   
...        ...      ...      ...           ...          ...            ...   
4995      4996       73      993           330            2          debit   
4996      4997       48      789           234            2           cash   
4997      4998       56      867           351            3           cash   
4998      4999       60      825           354            2    credit_card   
4999      5000       44      734           288            2          debit   

               created_at  
0     2017-03-13 12:36:56  
1     2017-03-03 17:38:52  
2      2017-03-14 4:23:56  
3     2017-03-26 12:43:37  
4      2017-03-01 4:35:11  
...                   ...  
4995  2017-03-30 13:47:17  
4996  2017-03-16 20:36:16  
4997   2017-03-19 5:42:42  
4998  2017-03-16 14:51:18  
4999  2017-03-18 15:48:18  

[4896 rows x 7 columns]

In [36]:
aov = orders_df["order_amount"].sum() / orders_df["total_items"].sum()
aov

372.7100853619981

###Q1c-What is its value?


###Answer -  As we can see above, the correct metric to evaluate the given dataset is AOV and the value is 372.71.

###Question 2: For this question you’ll need to use SQL. Follow this link to access the data set required for the challenge. Please use queries to answer the following questions. Paste your queries along with your final numerical answers below.


###Q2a - How many orders were shipped by Speedy Express in total?


###Answer -

SELECT count(*) as total_orders from Orders o
inner join Shippers s on o.ShipperId = s.ShipperId
where ShipperName = "Speedy Express"; 

(Answer- 54)

###Q2b - What is the last name of the employee with the most orders?

###Answer - 

SELECT LastName from Employees e
inner join Orders o on e.EmployeeId = o.EmployeeId
group by LastName
order by count(o.OrderId) desc LIMIT 1;

(Answer- Peacock)

###Q2c - What product was ordered the most by customers in Germany?

###Answer - 

SELECT p.ProductName, cus.Country, od.Quantity
FROM (((Orders o
INNER JOIN OrderDetails od ON
od.OrderID= o.OrderID)
INNER JOIN Products p ON
p.ProductID= od.ProductID)
INNER JOIN Customers cus ON
cus.CustomerID= o.CustomerID)
WHERE Country = "Germany"
GROUP BY p.ProductName, cus.Country, od.Quantity
order by Quantity DESC LIMIT 1;

(Answer - Steeleye Stout, Quatity: 100)